doc/ug/rust_for_c.md

```diff
@@ -0,0 +1,2340 @@
+Importantly, all of these actions require uttering the keyword `unsafe`, so that they can be easily detected in code review.
+Code inside unsafe blocks says to the  reader that the programmer has checked subtle safety guarantees that the compiler cannot on its own.
+
+Unsafe Rust fundamentally works by "turning off" certain checks the compiler normally enforces, whenever inside a `unsafe { /* ... */ }` block expression or the body of an `unsafe fn`.
+The things that Unsafe Rust can do that Safe Rust cannot are:
+*   Call `unsafe` functions.
+*   Dereference raw pointers.
+*   Mutate global state through a mutable `static`.
+*   Read union fields.
+*   Call the `asm!` macro.
+
+Additionally, `unsafe impl`s may implement `unsafe trait`s, but don't need to be inside an `unsafe` block.
+
+The canonical reference is the Rustonomicon[^127], a non-normative document describing common uses for Unsafe Rust.
+It is required reading (mostly the first half) for embedded programming.
+It contains detailed examples of correct and incorrect Unsafe Rust usage, and guidance on when to invoke Unsafe Rust.
+
+Throughout this document, references to Unsafe Rust are made, mostly around calling unsafe functions and referencing raw pointers, which are roughly all that Unsafe Rust can do that normal Rust can't.
+With these powers comes responsibility: Unsafe Rust is not safe from Undefined Behavior, and cannot leave the machine in a state where actions allowed in normal, safe Rust would trigger Undefined Behavior.
+In general, a few rules of thumb are useful:
+*   Every `unsafe fn` should declare, in documentation, what invariants it assumes that the caller will uphold, and what state it will leave the machine in.
+    For example, `<[T]>::get_unchecked(n)` elides the bounds check for the indexing operation, and it is up to the caller to uphold it instead.
+*   Every time `unsafe` code calls into a non-`unsafe` function, it must ensure that no violated invariants, which could trigger Undefined Behavior, are observable in that safe code.
+    For example, if we have a type that maintains the invariant that `len > 0`, and we temporarily set it to `0` during an `unsafe` block, it must be restored to `> 0` before any safe methods can be called on that type[^128].
+*   Unsafe code should be kept to the absolute minimum, and wrapped in safe interfaces that assert invariants, either through static type-system guarantees or through runtime checks.
+    Every line of unsafe code is a place where the engineering cost of Rust's guarantees are wasted.
+
+In other words, Safe Rust is able to freely assume that Rust's guarantees on aliasing, ownership, and representation of values hold at all times.
+This assumption is perversive: not only does the compiler use it aggressively optimize code for speed and size, but other library code, such as the destructors of wrapper types, assume it, too.
```

Comment:
Did you mean "pervasive"? "perversive" is an esoteric word choice here.

Reply:
I can't English.